File: algorithms/[1109]航班预订统计/solution.js
/**
 * @param {number[][]} bookings
 * @param {number} n
 * @return {number[]}
 */
const corpFlightBookings = function(bookings, n) {
  // 乘车问题 前缀和
  const res = new Array(n).fill(0);
  for (let i = 0; i < bookings.length; i += 1) {
    const [start, end, count] = bookings[i];
    res[start - 1] += count;
    if (end < n) {
      res[end] -= count;
    }
  }

  for (let i = 1; i < n; i += 1) {
    res[i] += res[i - 1];
  }

  return res;
};

File: algorithms/[1109]航班预订统计/solution.py
# topics = ["数组", "数学"]

from typing import List


class Solution:
    def corpFlightBookings(self, bookings: List[List[int]], n: int) -> List[int]:
        """
        差分数组，区间两端进行操作以代替区间内操作
        """
        res = [0] * n
        for i, j, cnt in bookings:
            res[i - 1] += cnt
            if j < n:
                res[j] -= cnt

        for i in range(1, n):
            res[i] += res[i - 1]

        return res
